File: backend/src/main/java/es/udc/tfgproject/backend/model/services/ReservationService.java
package es.udc.tfgproject.backend.model.services;

import java.math.BigDecimal;
import java.time.LocalDate;

import es.udc.tfgproject.backend.model.entities.EventEvaluation;
import es.udc.tfgproject.backend.model.entities.Menu;
import es.udc.tfgproject.backend.model.entities.Reserve;
import es.udc.tfgproject.backend.model.entities.Reserve.PeriodType;
import es.udc.tfgproject.backend.model.exceptions.CompanyDoesntAllowReservesException;
import es.udc.tfgproject.backend.model.exceptions.EmptyMenuException;
import es.udc.tfgproject.backend.model.exceptions.InstanceNotFoundException;
import es.udc.tfgproject.backend.model.exceptions.MaximumCapacityExceededException;
import es.udc.tfgproject.backend.model.exceptions.PermissionException;
import es.udc.tfgproject.backend.model.exceptions.ReservationDateIsBeforeNowException;

public interface ReservationService {

	Menu addToMenu(Long userId, Long menuId, Long productId, Long companyId, int quantity)
			throws InstanceNotFoundException, PermissionException;

	Menu updateMenuItemQuantity(Long userId, Long menuId, Long productId, Long companyId, int quantity)
			throws InstanceNotFoundException, PermissionException;

	Menu removeMenuItem(Long userId, Long menuId, Long productId, Long companyId)
			throws InstanceNotFoundException, PermissionException;

	Menu findMenuProducts(Long userId, Long menuId, Long companyId)
			throws InstanceNotFoundException, PermissionException;

	Reserve reservation(Long userId, Long menuId, Long companyId, LocalDate reservationDate, Integer diners,
			PeriodType periodType) throws InstanceNotFoundException, PermissionException, EmptyMenuException,
			MaximumCapacityExceededException, ReservationDateIsBeforeNowException, CompanyDoesntAllowReservesException;

	void cancelReservation(Long userId, Long reserveId) throws InstanceNotFoundException, PermissionException;

	void removeReservation(Long userId, Long reserveId) throws InstanceNotFoundException, PermissionException;

	BigDecimal calculateDepositFromPercentage(Long companyId, BigDecimal totalPrice) throws InstanceNotFoundException;

	Boolean checkCapacity(Long companyId, LocalDate reservationDate, PeriodType periodType, Integer diners)
			throws MaximumCapacityExceededException, InstanceNotFoundException, ReservationDateIsBeforeNowException;

	Integer obtainMaxDinersAllowed(Long companyId, LocalDate reservationDate, PeriodType periodType)
			throws InstanceNotFoundException;

	Reserve findReserve(Long userId, Long reserveId) throws InstanceNotFoundException, PermissionException;

	Block<Reserve> findUserReserves(Long userId, int page, int size);

	Block<Reserve> findCompanyReserves(Long userId, Long companyId, LocalDate date, PeriodType periodType, int page,
			int size) throws InstanceNotFoundException, PermissionException;

	Block<Reserve> findCompanyReservesCanceled(Long userId, Long companyId, int page, int size)
			throws InstanceNotFoundException, PermissionException;

	void addEventEvaluation(Long userId, Long eventEvaluationId, Integer points, String opinion)
			throws PermissionException, InstanceNotFoundException;

	Block<EventEvaluation> findUserEventEvaluations(Long userId, int page, int size)
			throws PermissionException, InstanceNotFoundException;

	Block<EventEvaluation> findCompanyEventEvaluations(Long userId, Long companyId, int page, int size)
			throws PermissionException, InstanceNotFoundException;

}

File: frontend/src/backend/reservationService.js
import {config, appFetch} from './appFetch';

export const addToMenu = (menuId, productId, companyId, quantity, onSuccess, 
    onErrors) =>
    appFetch(`/reservation/menus/${menuId}/addToMenu`, 
        config('POST', {productId, companyId, quantity}), onSuccess, onErrors);

export const updateMenuItemQuantity = (menuId, productId, companyId,
    quantity, onSuccess, onErrors) =>
    appFetch(`/reservation/menus/${menuId}/updateMenuItemQuantity`, 
        config('POST', {productId, companyId, quantity}), onSuccess, onErrors);

export const removeMenuItem = (menuId, productId, companyId, onSuccess,
    onErrors) =>
    appFetch(`/reservation/menus/${menuId}/removeMenuItem`, 
        config('POST', {productId, companyId}), onSuccess, onErrors);

export const findMenuProducts = (menuId, companyId,  onSuccess) => 
    appFetch(`/reservation/menus/${menuId}?companyId=${companyId}`, 
        config('GET'), onSuccess);

export const reservation = (menuId, companyId, reservationDate, diners, periodType, saleId,
    onSuccess, onErrors) =>
    appFetch(`/reservation/menus/${menuId}/reservation`, 
        config('POST', {companyId, reservationDate, diners, periodType, saleId}), onSuccess, onErrors);

export const cancelReservation = (reserveId, onSuccess, onErrors) =>
    appFetch(`/reservation/reserves/${reserveId}`,
        config('DELETE'), onSuccess, onErrors);

export const calculateDepositFromPercentage = (companyId, totalPrice, onSuccess) => 
    appFetch(`/reservation/menus/calculateDeposit?companyId=${companyId}&totalPrice=${totalPrice}`, 
        config('GET'), onSuccess);

export const checkCapacity = (companyId, reservationDate, periodType, diners, onSuccess, onErrors) => 
    appFetch(`/reservation/menus/checkCapacity?companyId=${companyId}&reservationDate=${reservationDate}
        &periodType=${periodType}&diners=${diners}`, 
        config('GET'), onSuccess, onErrors);
        
export const obtainMaxDinersAllowed = (companyId, reservationDate, periodType, onSuccess) => 
    appFetch(`/reservation/menus/obtainMaxDinersAllowed?companyId=${companyId}&reservationDate=${reservationDate}
        &periodType=${periodType}`, 
        config('GET'), onSuccess);

export const findReserve = (reserveId, onSuccess) =>
    appFetch(`/reservation/reserves/${reserveId}`, config('GET'), onSuccess);

export const findUserReserves = ({page}, onSuccess) => 
    appFetch(`/reservation/userReserves?page=${page}`, config('GET'), onSuccess);

export const findCompanyReserves = (companyId, reservationDate, periodType, {page}, onSuccess) => 
    appFetch(`/reservation/companyReserves?companyId=${companyId}&reservationDate=${reservationDate}
        &periodType=${periodType}&page=${page}`, config('GET'), onSuccess);

export const findCompanyReservesCanceled = (companyId, {page}, onSuccess) => 
    appFetch(`/reservation/companyReservesCanceled?companyId=${companyId}&page=${page}`, 
    config('GET'), onSuccess);

export const addEventEvaluation = (eventEvaluationId, points, opinion,
    onSuccess, onErrors) =>
    appFetch(`/reservation/eventEvaluation/${eventEvaluationId}`, 
        config('POST', {eventEvaluationId, points, opinion}), onSuccess, onErrors);

export const findUserEventEvaluations = ({page}, onSuccess) => 
    appFetch(`/reservation/userEventEvaluations?page=${page}`, config('GET'), onSuccess);

export const findCompanyEventEvaluations = (companyId, {page}, onSuccess) => 
    appFetch(`/reservation/companyEventEvaluations?companyId=${companyId}&page=${page}`, 
        config('GET'), onSuccess);
        